toucan_connectors/http_api/http_api_connector.py

```diff
@@ -183,24 +115,63 @@ def do_request(self, query, session):
                         f'Cannot decode response content from query: method={query.get("method")} url={query.get("url")} response_status_code={res.status_code} response_reason=${res.reason}'  # noqa: E501
                     )
                     raise
+        return data
+
+    def perform_requests(
+        self, data_source: HttpAPIDataSource, session: Session, pagination_config: PaginationConfig
+    ) -> list[Any]:
+        data_source = pagination_config.apply_pagination_to_data_source(data_source)
+        query = self._render_query(data_source)
+        jq_filter = query["filter"]
+        raw_result = self.do_request(query, session)
+
         try:
-            return transform_with_jq(data, jq_filter)
+            parsed_result = transform_with_jq(raw_result, jq_filter)
+            parsed_pagination_info = None
+            if jq_pagination_filter := pagination_config.get_pagination_info_filter():
+                # transform_with_jq search for all occurrences, it will always return a list of values
+                parsed_pagination_info = transform_with_jq(raw_result, jq_pagination_filter)[0]
         except ValueError:
-            HttpAPIConnector.logger.error(f"Could not transform {data} using {jq_filter}")
+            HttpAPIConnector.logger.error(f"Could not transform {raw_result} using {jq_filter}")
             raise
 
+        results = []
+        if next_pagination_config := pagination_config.get_next_pagination_config(
+            result=parsed_result, pagination_info=parsed_pagination_info
+        ):
+            results += self.perform_requests(
+                data_source=data_source, session=session, pagination_config=next_pagination_config
+            )
+
+        results += parsed_result
+        return results
+
     def _retrieve_data(self, data_source: HttpAPIDataSource) -> "pd.DataFrame":
         if self.auth:
             session = self.auth.get_session()
         else:
             session = Session()
 
-        query = self._render_query(data_source)
+        try:
+            results = pd.DataFrame(
+                self.perform_requests(
+                    data_source=data_source,
+                    session=session,
+                    pagination_config=self.http_pagination.get_pagination_config(),
+                )
+            )
+        except HTTPError as exc:
+            if exc.response.status_code == TOO_MANY_REQUESTS:
+                raise HTTPError(
```

Comment:
A custom exception should be raised here, as you're removing information from the original information, such as the status code: If some other code is catching `HTTPError` higher up, if might have unexpected behavior since some information it expects to find in the exception wouldn't be present